File: src/db/models/view_history.js
module.exports = function(sequelize, DataTypes) {
    var ViewHistory = sequelize.define(
        'ViewHistory',
        {
            id: {
                allowNull: false,
                autoIncrement: true,
                primaryKey: true,
                type: DataTypes.INTEGER,
            },
            UserId: {
                type: DataTypes.INTEGER,
                references: {
                    model: 'users',
                    key: 'id',
                },
                field: 'user_id',
            },
            ContentId: {
                type: DataTypes.INTEGER,
                references: {
                    model: 'contents',
                    key: 'id',
                },
                field: 'content_id',
            },
            residence_time: {
                type: DataTypes.DATE,
                defaultValue: new Date(),
            },
            meta: {
                type: DataTypes.STRING(255),
                defaultValue: '',
            },
        },
        {
            tableName: 'viewHistories',
            createdAt: 'created_at',
            updatedAt: 'updated_at',
            timestamps: true,
            underscored: true,
            classMethods: {
                associate: function(models) {
                    ViewHistory.belongsTo(models.User, {
                        onDelete: 'CASCADE',
                        foreignKey: {
                            name: 'user_id',
                            allowNull: false,
                        },
                    });
                    ViewHistory.belongsTo(models.Content, {
                        onDelete: 'CASCADE',
                        foreignKey: {
                            name: 'content_id',
                            allowNull: false,
                        },
                    });
                },
            },
        }
    );

    ViewHistory.Instance.prototype.toJSON = self => {
        return {
            id: self.id,
            UserId: self.UserId,
            ContentId: self.ContentId,
            residence_time: self.residence_time,
            meta: self.meta,
        };
    };

    return ViewHistory;
};

File: src/python-scripts/models/view_history.py
import re
from collections import OrderedDict
import sys, json
import numpy as np
from models.content import Contents, Content
from models.user import Users, User

class ViewHistories(object):
    def __init__(self, JSON):
        self.items = [ViewHistory(val) for val in JSON if val is not None]

    def toSortArrayWithContent(self):
        return { val.ContentId : [ item.UserId for item in self.getsByContentId(val.ContentId) ] for val in self.items }

    def toJSON(self):
        return { "viewHistories": [val.toJSON() for val in self.items] }

    def getById(self, id):
        for item in self.items:
            if item.id == id:
                return item
            continue

    def getsByContentId(self, id):
        return [item for item in self.items if item.ContentId == id]


class ViewHistory(object):
    def __init__(self, JSON):
        if JSON is None:
            return
        self.id = JSON.get('id')
        self.UserId = JSON.get('UserId')
        self.ContentId = JSON.get('ContentId')
        self.meta = JSON.get('meta')
        self.isPrivate = JSON.get('isPrivate')
        self.valid = JSON.get('valid')
        self.permission = JSON.get('permission')
        # self.Content = Content(JSON.get("Content"))

    def toJSON(self):
        return {
            'id': self.id,
            'UserId': self.UserId,
            'ContentId': self.ContentId,
            'meta': self.meta,
            'isPrivate': self.isPrivate,
            'valid': self.valid,
            'permission': self.permission,
        }
